File: slf4j/src/test/kotlin/SimpleTest.kt
import notion.api.v1.NotionClient
import notion.api.v1.http.OkHttp3Client
import notion.api.v1.logging.Slf4jLogger
import notion.api.v1.model.pages.Page
import notion.api.v1.model.pages.PageProperty
import notion.api.v1.request.pages.CreatePageRequest
import notion.api.v1.request.pages.UpdatePagePropertiesRequest
import org.junit.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class SimpleTest {

    @Test
    fun pages() {
        NotionClient(token = System.getenv("NOTION_TOKEN")).use { client ->
            client.httpClient = OkHttp3Client()
            client.logger = Slf4jLogger()

            val databases = client.listDatabases()
            assertTrue { databases.results.isNotEmpty() }

            val database = databases.results[0]

            val newPage: Page = client.createPage(CreatePageRequest(
                parent = CreatePageRequest.Parent(
                    type = "database",
                    databaseId = database.id,
                ),
                properties = mapOf(
                    "Title" to PageProperty(
                        title = listOf(
                            PageProperty.RichText(
                                text = PageProperty.RichText.Text(content = "Fix a bug")
                            )
                        )
                    ),
                    "Description" to PageProperty(
                        richText = listOf(
                            PageProperty.RichText(
                                text = PageProperty.RichText.Text(
                                    content = "The bug is a minor one but one of our major customers have been affected by it for a while."
                                )
                            )
                        )
                    ),
                    "Due" to PageProperty(
                        date = PageProperty.Date(start = "2021-05-13", end = "2021-12-31")
                    ),
                    "Severity" to PageProperty(
                        select = database.properties["Severity"]?.select?.options?.find { it.name == "High" }
                    ),
                )
            ))
            assertNotNull(newPage)

            val patchResult = client.updatePageProperties(UpdatePagePropertiesRequest(
                pageId = newPage.id,
                properties = mapOf(
                    "Due" to PageProperty(
                        date = PageProperty.Date(start = "2021-12-31")
                    ),
                    "Severity" to PageProperty(
                        select = database.properties["Severity"]?.select?.options?.find { it.name == "Medium" }
                    ),
                )
            ))
            assertNotEquals(newPage.lastEditedTime, patchResult.lastEditedTime)

            val fetchedPage = client.retrievePage(newPage.id)
            assertEquals(patchResult.lastEditedTime, fetchedPage.lastEditedTime)
        }
    }
}

File: scala3/src/test/scala/SimpleTest.scala
import notion.api.v1.ScalaNotionClient
import notion.api.v1.http.JavaNetHttpClient
import notion.api.v1.model.databases.query.filter.condition.TextFilter
import notion.api.v1.model.databases.query.filter.PropertyFilter
import notion.api.v1.model.pages.PageProperty
import notion.api.v1.model.pages.PageProperty.RichText
import notion.api.v1.model.pages.PageProperty.RichText.Text
import notion.api.v1.request.pages.CreatePageRequest.Parent
import org.junit.Assert.assertNotNull
import org.junit.Test

import scala.jdk.CollectionConverters._

class SimpleTest {

  @Test
  def databases(): Unit = {
    val client = ScalaNotionClient(System.getenv("NOTION_TOKEN"))
    client.httpClient = new JavaNetHttpClient()
    try {
      val databases = client.listDatabases()
      assertNotNull(databases)

      val databaseId = databases.getResults.asScala
        .filter(_.getTitle.get(0).getPlainText == "Test Database")
        .head
        .getId

      val database = client.retrieveDatabase(databaseId)
      assertNotNull(database)

      val queryResult =
        client.queryDatabase(
          databaseId,
          filter = {
            val f = new PropertyFilter()
            f.setProperty("Title")
            val title = new TextFilter()
            title.setContains("bug")
            f.setTitle(title)
            f
          },
          pageSize = 1
        )
      assertNotNull(queryResult)
    } finally client.close()
  }

  @Test
  def pages(): Unit = {
    val client = ScalaNotionClient(
      token = System.getenv("NOTION_TOKEN"),
      httpClient = new JavaNetHttpClient()
    )
    try {
      val databases = client.listDatabases()
      assertNotNull(databases)

      val databaseId = databases.getResults.asScala
        .filter(_.getTitle.get(0).getPlainText == "Test Database")
        .head
        .getId

      val page = client.createPage(
        parent = {
          val p = new Parent("database")
          p.setDatabaseId(databaseId)
          p
        },
        properties = Map(
          "Title" -> {
            val p = new PageProperty()
            p.setTitle(List({
              val rt = new RichText()
              rt.setText(new Text("foo", null))
              rt
            }).asJava)
            p
          }
        )
      )
      assertNotNull(page)

      val updatedPage = client.updatePageProperties(
        pageId = page.getId,
        properties = Map(
          "Title" -> {
            val p = new PageProperty()
            p.setTitle(List({
              val rt = new RichText()
              rt.setText(new Text("This is an important task!", null))
              rt
            }).asJava)
            p
          }
        )
      )
      assertNotNull(updatedPage)

      val fetchedPage = client.retrievePage(page.getId)
      assertNotNull(fetchedPage)

    } finally client.close()
  }

}
